src/meta/src/rpc/leader_svs.rs

```diff
@@ -0,0 +1,316 @@
+use risingwave_common_service::metrics_manager::MetricsManager;
+use risingwave_object_store::object::object_metrics::ObjectStoreMetrics;
+use risingwave_object_store::object::parse_remote_object_store;
+use risingwave_pb::meta::MetaLeaderInfo;
+use tokio::sync::oneshot::{Receiver as OneReceiver, Sender as OneSender};
+use tokio::sync::watch::Receiver as WatchReceiver;
+use tokio::task::JoinHandle;
+
+use super::service::health_service::HealthServiceImpl;
+use super::service::notification_service::NotificationServiceImpl;
+use super::service::scale_service::ScaleServiceImpl;
+use super::DdlServiceImpl;
+use crate::backup_restore::BackupManager;
+use crate::barrier::{BarrierScheduler, GlobalBarrierManager};
+use crate::hummock;
+use crate::hummock::{CompactionScheduler, HummockManager};
+use crate::manager::{
+    CatalogManager, ClusterManager, FragmentManager, IdleManager, MetaOpts, MetaSrvEnv,
+};
+use crate::rpc::metrics::MetaMetrics;
+use crate::rpc::server::AddressInfo;
+use crate::rpc::service::backup_service::BackupServiceImpl;
+use crate::rpc::service::cluster_service::ClusterServiceImpl;
+use crate::rpc::service::heartbeat_service::HeartbeatServiceImpl;
+use crate::rpc::service::hummock_service::HummockServiceImpl;
+use crate::rpc::service::stream_service::StreamServiceImpl;
+use crate::rpc::service::user_service::UserServiceImpl;
+use crate::storage::MetaStore;
+use crate::stream::{GlobalStreamManager, SourceManager};
+
+/// Simple struct containing everything required to start services on a meta leader node
+pub struct LeaderServices<S: MetaStore> {
+    pub backup_srv: BackupServiceImpl<S>,
+    pub cluster_srv: ClusterServiceImpl<S>,
+    pub ddl_srv: DdlServiceImpl<S>,
+    pub health_srv: HealthServiceImpl,
+    pub heartbeat_srv: HeartbeatServiceImpl<S>,
+    pub hummock_srv: HummockServiceImpl<S>,
+
+    /// OneShot receiver that signals if sub tasks is idle
+    pub idle_recv: OneReceiver<()>,
+
+    pub meta_metrics: Arc<MetaMetrics>,
+    pub notification_srv: NotificationServiceImpl<S>,
+    pub scale_srv: ScaleServiceImpl<S>,
+    pub stream_srv: StreamServiceImpl<S>,
+
+    /// sub_tasks executed concurrently. Can be shutdown via shutdown_all
+    pub sub_tasks: Vec<(JoinHandle<()>, OneSender<()>)>,
+
+    pub user_srv: UserServiceImpl<S>,
+}
+
+/// Initializing all services needed for the meta leader node
+/// Only call this function once, since initializing the services multiple times will result in an
+/// inconsistent state
+///
+/// ## Returns
+/// Returns an instance of `LeaderServices` or an Error if initializing leader services failed
+pub async fn get_leader_srv<S: MetaStore>(
+    meta_store: Arc<S>,
+    address_info: AddressInfo,
+    max_heartbeat_interval: Duration,
+    opts: MetaOpts,
+    leader_rx: WatchReceiver<(MetaLeaderInfo, bool)>,
+    election_handle: JoinHandle<()>,
+    election_shutdown: OneSender<()>,
+) -> Result<LeaderServices<S>, BackupError> {
```

Comment:
Why it's a `BackupError`? Can we use `MetaResult` here?

Reply:
Sure 👍